### PR TITLE
Add a draft for a how-to that explains setting charm/workload versions

### PR DESCRIPTION
**Pull request by tonyandrewmeyer**
For adding as a new "How to"/"Develop"/"Set charm and workload versions" guide.

Also with the intention that when the how-to guide exists, [chapter 2 of the k8s tutorial](https://juju.is/docs/sdk/expose-the-version-of-the-application-behind-your-charm) isn't really needed any more and could be removed to make a leaner tutorial.

### CODE REVIEW

**Comment by tonyandrewmeyer** [comment]
@tmihoc could you please review this when you have some time available? Thanks!

File: sdk/how-to-guides/versions.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+```shell
+$ git rev-parse HEAD > version
+$ ls
+lib      src    tox.ini charmcraft.yaml  LICENSE  requirements.txt  tests  version
+$ cat version
+0522e1fd009dac78adb3d0652d91a1e8ff7982ae
+```
+
+<note>
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: Here and at the end: Change < to square brackets. (It's the correct notation for notes on Discourse.)

File: sdk/how-to-guides/versions.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+- [Set the workload version](#heading--set-the-version-of-the-workload)
+- [Test setting the workload version](#heading--test-setting-the-workload-version)
+
+<a href="#heading--set-the-version-of-the-charm"><h2 id="heading--set-the-version-of-the-charm">Set the charm version</h3></a>
+
+Charms can specify a version of the charm itself, to allow tracking the
+installed version of the charm back to the source tree that it was built from.
+
+To set the version, include a file called `version` (no extension) at the top
+level of the charm (the same folder as the `charmcraft.yaml` file). The content
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: suggestion: Rephrase as: 

> To set the version, in the root directory of your charm (at the same level as the `charmcraft.yaml` file) add a file called `version`.

rationale: This order better matches the order in which people will have to execute things (first find the root directory, then add this file). Also, "include" may start people on other mental paths (as in code includes) -- it'll be easy for them to course-correct, but why start them on those paths to begin with?

File: sdk/how-to-guides/versions.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
+**Contents:**
+
+- [Set the charm version](#heading--set-the-version-of-the-charm)
+- [Test setting the charm version](#heading--test-setting-the-charm-version)
+- [Set the workload version](#heading--set-the-version-of-the-workload)
+- [Test setting the workload version](#heading--test-setting-the-workload-version)
+
+<a href="#heading--set-the-version-of-the-charm"><h2 id="heading--set-the-version-of-the-charm">Set the charm version</h3></a>
+
+Charms can specify a version of the charm itself, to allow tracking the
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: suggestion: Make the "to allow tracking..." clearer. Where does this become possible and to whom? I see you clarify that below, but maybe you can give a hint here already: "to allow Juju to track the installed..."

File: sdk/how-to-guides/versions.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+installed version of the charm back to the source tree that it was built from.
+
+To set the version, include a file called `version` (no extension) at the top
+level of the charm (the same folder as the `charmcraft.yaml` file). The content
+of the file is the version string, which is typically a major.minor.patch style
+verison or a version control hash identifier.
+
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: typo: version*

Also, for the major.minor.patch case, do they just insert it manually? As with all the how-to guides, it would be good to make the generic recipe as clear as possible in the abstract, then give an example  (you do that for the hash), then link to real-world examples (you could add some at the end of the section). Also, is there a preference for one over the other?

File: sdk/how-to-guides/versions.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+> See more: [Juju | Revision](https://juju.is/docs/sdk/revision)
+
+<a href="#heading--test-setting-the-charm-version"><h2 id="heading--test-setting-the-charm-version">Test setting the charm version</h2></a>
+
+> See first: [Get started with charm testing](https://juju.is/docs/sdk/get-started-with-charm-testing)
+
+Since the version isn't set by the charm code itself, you'll want to test that
+the version is correctly set with an integration test, rather than also having a
+unit test.
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: suggestion: Rephrase "rather than also having" without "also" (I'm finding it confusing). 

PS "Since the version isn't set by the charm code... integration test" -- this felt insightful. Wondering why it wasn't crystal clear to me already. Will check our content on integration testing again.

File: sdk/how-to-guides/versions.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+> See first: [How to write integration tests for a charm](/t/12734)
+
+To verify that setting the charm version works correctly in an integration test, get the status
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: suggestion: After "in an integration test," insert: in your `charm.py`, in the body of..., get the status...

rationale: It may feel superfluous but it will make each how-to guide a lot more self-sufficient down the line, which is always a good thing.

File: sdk/how-to-guides/versions.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    status = await ops_test.model.get_status()
+    version = status.applications[APP_NAME].charm_version
+    expected_version = subprocess.check_output(["git", "rev-parse", "HEAD"]).decode("utf8")
+    assert version == expected_version
+```
+
+<!---
+No "see more" link: this is not currently documented in the pylibjuju docs.
+-->
+
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: Isn't this pytest-operator material? I know it uses pylibjuju in the background, but I also know it has some docs of its own.

File: sdk/how-to-guides/versions.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+<a href="#heading--set-the-version-of-the-workload"><h2 id="heading--set-the-version-of-the-workload">Set the workload version</h3></a>
+
+Most applications modelled by charms will also have a version. Each application
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: "most" = ? (Which ones do not?)

File: sdk/how-to-guides/versions.md [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+will have its own versioning scheme, and its own way of accessing that
+information. To make things easier for Juju admins, the charm should expose the
+workload version through Juju - it's visible in `juju status` in the application
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: suggestion: "-- it will be visible in `juju status`  (in the default tabular view, in the application table > the "Version column"; in the JSON or YAML format, under `applications.<app name>.version`).

File: sdk/how-to-guides/versions.md [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+table in the "Version" column in the tabular view, or in the
+`applications.<app name>.version` field in JSON or YAML format.
+
+<note>
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: Change to square brackets.

File: sdk/how-to-guides/versions.md [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+```python
+def _on_start(self, event: ops.StartEvent):
+    # The workload exposes the version via HTTP at /version
+    version = requests.get("http://localhost:8000/version").text
+    self.unit.set_workload_version(version)
+```
+
+> See more: [`ops.Unit.set_workload_version`](https://ops.readthedocs.io/en/latest/#ops.Unit.set_workload_version)
+
+> Examples: [jenkins-k8s sets the workload version after getting it from the Jenkins package](https://github.com/canonical/jenkins-k8s-operator/blob/29e9b652714bd8314198965c41a60f5755dd381c/src/charm.py#L115), [discourse-k8s sets the workload version after getting it via an exec call](https://github.com/canonical/discourse-k8s-operator/blob/f523b29f909c69da7b9510b581dfcc2309698222/src/charm.py#L581), [synapse sets the workload version after getting it via an API call](https://github.com/canonical/synapse-operator/blob/778bcd414644c922373d542a304be14866835516/src/charm.py#L265)
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: I'd add backticks around the charm names. 

(BTW Your comments on what sort of example the link is going to show are really good.)

File: sdk/how-to-guides/versions.md [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+
+<a href="#heading--test-setting-the-workload-version"><h2 id="heading--test-setting-the-workload-version">Test setting the workload version</h2></a>
+
+> See first: [Get started with charm testing](https://juju.is/docs/sdk/get-started-with-charm-testing)
+
+You'll want to add three levels of tests:
+
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: This list of skip-ahead links is very helpful. However, I don't like how it's currently functioning as both part of the running text (it completes the previous sentence) and meta text (TOC). I'd keep it as just a TOC and revise the intro sentence to be complete on its own:

> You'll want to add three levels of tests: unit, scenario, and integration.

File: sdk/how-to-guides/versions.md [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+- [Write unit tests](#heading--write-unit-tests)
+- [Write scenario tests](#heading--write-scenario-tests)
+- [Write integration tests](#heading--write-integration-tests)
+
+<a href="#heading--write-unit-tests"><h3 id="heading--write-unit-tests">Write unit tests</h3></a>
+
+> See first: [How to write unit tests for a charm](/t/4461)
+
+When using Harness for unit tests, use the `get_workload_version()` method to
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: "When.." suggests there are other ways. I see below that you call scenario tests another type of unit tests. I guess that makes sense -- but in that case the mini-TOC hierarchy needs to have just two main titles, unit and integration. Also, before we make that change, another point to consider: charmers tend to think of scenario tests as their own kind and typically keep them separate under "scenario tests". 

Whichever way we go, I'd rephrase as: To [e.g, Harness-unit-test] the workload version, in your [clarify where the code needs to go] use the `ops.harness.get_workoad_version()` method. For example, below we check if the version is 1.2: ...

On being explicit about the where: In my experience not saying where things need to happen is a common omission in documentation, one that hurts a lot (especially beginners but can be useful to experienced users too), and one that is however very easily fixed. As a result, I always recommend being explicit not just about what needs to happen but also where it needs to happen. Note that I'm not trying to blur the lines between a tutorial and a how-to guide here -- just trying to make the how-to guide as accessible and usable as it can be while remaining compact and to the point.

File: sdk/how-to-guides/versions.md [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+
+<note>
+If the charm has not set the workload version, then the field will not be
+present in JSON or YAML format, and if the version string is too long or
+contains particular characters then it will not be displayed in the tabular
+format.
+</note>
+
+The workload version is often set in one of the early event handlers, such as
+`start`. If you don't already have a `start` handler, in the `src/charm.py`
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: What other handlers might make sense as well and why?

File: sdk/how-to-guides/versions.md [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+<a href="#heading--test-setting-the-workload-version"><h2 id="heading--test-setting-the-workload-version">Test setting the workload version</h2></a>
+
+> See first: [Get started with charm testing](https://juju.is/docs/sdk/get-started-with-charm-testing)
+
+You'll want to add three levels of tests:
+
+- [Write unit tests](#heading--write-unit-tests)
+- [Write scenario tests](#heading--write-scenario-tests)
+- [Write integration tests](#heading--write-integration-tests)
+
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: The titles sound generic although your goal here is to write tests specifically in the context of testing the worklod version. I'd update them that way. E.g.,

- Test setting the workload version -- unit tests
- Test setting the workload version -- scenario tests
- Test setting the workload version -- integration tests

I wouldn't worry about the titles feeling redundant -- they make for good SEO and also provide a lot more context locally so overall they should just contribute to better clarity and focus.

If you can think of a way tomake them less clunky, that would however be nice. Maybe "Unit-test setting the workload version", "Scenario-test...", "Integration-test..."?

File: sdk/how-to-guides/versions.md [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+def test_start(harness):
+    # Our charm gets the workload version by running a command in the container,
+    # so mock that out.
+    harness.handle_exec("webserver", ["/bin/server", "--version"], result="1.2\n")
+    # begin_with_initial_hooks will trigger the 'start' event, and we expect
+    # the charm's 'start' handler to set the workload version.
+    harness.begin_with_initial_hooks()
+    assert harness.get_workload_version() == "1.2"
+```
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: The comments here are very helpful -- thanks! However, which charm does "our charm gets..." specifically refer to? If it's an example started above (which I can't quite see at the moment), that makes the how-to guide entangled. I think it's better if everything about it stays quite generic. One way to achieve that here would be to rephrase as:

> Suppose your charm is a Kubernetes charm that gets...

File: sdk/how-to-guides/versions.md [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+    # version. If the workload has an API or some other way of getting the
+    # version, the test should get it from there and use that to compare to the
+    # unit setting.
+    assert version == "3.14"
+```
+
+<!---
+No "see more" link: this is not currently documented in the pylibjuju docs.
+-->
+
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: suggestion: Create an issue under pylibjuju to request documentation. Though my understand was for integration testing you use pylibjuju _through pytest_operator_, so then the issue needs to be filed _there_ instead?

File: sdk/how-to-guides/versions.md [hunk 18]
```diff
@@ -0,0 +1,216 @@
+**Contents:**
+
+- [Set the charm version](#heading--set-the-version-of-the-charm)
+- [Test setting the charm version](#heading--test-setting-the-charm-version)
+- [Set the workload version](#heading--set-the-version-of-the-workload)
+- [Test setting the workload version](#heading--test-setting-the-workload-version)
+
```
Comment (tmihoc) [inline sdk/how-to-guides/versions.md]: (1) What would be your suggestion for the title of the entire doc, given these contents? 

(2) The TOC feels like it's need more hierarchy, as the first two titles and the last two each form a story -- the charm version and the workload version story, respectively. 

(3) How does this doc connect to https://juju.is/docs/sdk/workloads and https://juju.is/docs/sdk/interact-with-pebble ? (We don't have to solve this problem now but it feels like a discussion worth having because the workload bit feels like it belongs in the latter docs, but those docs have issues too, and we don't have much on the charm itself from a developer perspective -- that is, nothing beyond the charmcraft side of things.)

**Review by tmihoc — CHANGES_REQUESTED** [review]
Thank you so much for your work on this -- it's turned out beautiful! Left some comments, but I think only the small ones are for now -- the bigger questions (about relations to other how-to guides we have) are food for thought for another iteration.

File: sdk/how-to-guides/versions.md (hunk 17 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/versions.md]: Opened: https://github.com/juju/python-libjuju/issues/1096

This is straight pylibjuju, not anything that pytest-operator is adding.

File: sdk/how-to-guides/versions.md (hunk 7 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/versions.md]: We covered this when we talked, but for reference: the [pytest-operator reference docs](https://github.com/charmed-kubernetes/pytest-operator/blob/main/docs/reference.md) only cover the bits that pytest-operator adds (that's mostly convenience wrappers for building and deploying on a temporary model). The bulk of what people write, outside of the boilerplate, for integration tests is directly pylibjuju.

File: sdk/how-to-guides/versions.md (hunk 18 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/versions.md]: > (1) What would be your suggestion for the title of the entire doc, given these contents?

What about:

How to > Develop > Set the charm and application version

I'm not sure whether "application version" or "workload version" is better. I think we use the terms fairly interchageably - do you prefer one over the other? The Juju attribute is `workload_version`, but I think people generally understand "application" more than "workload".

> (2) The TOC feels like it's need more hierarchy, as the first two titles and the last two each form a story -- the charm version and the workload version story, respectively.

Yes, there really are two stories here. Another option, if short how-tos are fine, would be that these are two separate docs:

How to > Develop > Set the charm version
How to > Develop > Set the application version

Otherwise, I guess I could just put "Charm version" and "Application version" headings in?

> (3) How does this doc connect to https://juju.is/docs/sdk/workloads and https://juju.is/docs/sdk/interact-with-pebble ?

The charm version part doesn't (which is perhaps another reason it should be split off.

For the workload/application version, it does somewhat relate, but setting the version is the same whether you're writing a machine or k8s charm (except for the event that you're probably going to do it in), and I wouldn't want to repeat the content in both.

File: sdk/how-to-guides/versions.md (hunk 13 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/versions.md]: > "When.." suggests there are other ways. I see below that you call scenario tests another type of unit tests. I guess that makes sense -- but in that case the mini-TOC hierarchy needs to have just two main titles, unit and integration. Also, before we make that change, another point to consider: charmers tend to think of scenario tests as their own kind and typically keep them separate under "scenario tests".

That's a fair call. I guess I'm trying to push them to not think of them like that and think of them as the new/better way to write unit tests. But I'm probably getting ahead of myself and should leave that until Scenario is fully baked into ops and do it then.

**Comment by tonyandrewmeyer** [comment]
@tmihoc I've adjusted based on your feedback, thanks! Feel free to re-review, or otherwise just tell me to publish this and it can be refined later as per the usual processes.

Although we do need to decide if this is one HTG or two.

File: sdk/how-to-guides/versions.md (hunk 18 above)
Comment (tmihoc) [reply sdk/how-to-guides/versions.md]: Although they're both about versions, I feel like the two are unrelated (please correct me if I'm wrong). In that sense, let's have two docs. The placement you suggested under Develop makes sense. (I'm still not happy with the macro-level architecture in the SDK docs, but it's orthogonal to the present issue. This is good work that will help us think better about that issue as well.)

**Comment by tmihoc** [comment]
Replied on the other comment: Let's have two docs, and thanks!!
